synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -932,3 +937,77 @@ def _calculate_chain_cover_txn(
             processed_count=count,
             finished_room_map=finished_rooms,
         )
+
+    async def _purged_chain_cover_index(self, progress: dict, batch_size: int) -> int:
+        """
+        A background updates that iterates over the chain cover and deletes the
+        chain cover for events that have been purged.
+
+        This may be due to fully purging a room or via setting a retention policy.
+        """
+        current_event_id = progress.get("current_event_id", "")
+
+        def purged_chain_cover_txn(txn) -> int:
+            # The event ID from events will be null if the chain ID / sequence
+            # number points to a purged event.
+            sql = """
+                SELECT event_id, chain_id, sequence_number, e.event_id
+                FROM event_auth_chains
+                LEFT JOIN events AS e USING (event_id)
+                WHERE event_id > ? ORDER BY event_auth_chains.event_id ASC LIMIT ?
+            """
+            txn.execute(sql, (current_event_id, batch_size))
+
+            rows = txn.fetchall()
+            if not rows:
+                return 0
+
+            # The event IDs and chain IDs / sequence numbers where the event has
+            # been purged.
+            unreferenced_event_ids = []
+            unreferenced_chain_id_tuples = []
+            event_id = ""
+            for event_id, chain_id, sequence_number, events_event_id in rows:
+                if not events_event_id:
+                    unreferenced_event_ids.append(event_id)
+                    unreferenced_chain_id_tuples.append((chain_id, sequence_number))
+
+            # Delete the unreferenced auth chains from event_auth_chain_links and
+            # event_auth_chains.
+            txn.executemany(
+                """
+                DELETE FROM event_auth_chains WHERE event_id = ?
+                """,
+                unreferenced_event_ids,
+            )
+            txn.executemany(
+                """
+                DELETE FROM event_auth_chain_links WHERE
+                (origin_chain_id = ? AND origin_sequence_number = ?) OR
+                (target_chain_id = ? AND target_sequence_number = ?)
```

Comment:
We don't have an index on `target_chain_id` annoyingly. Since this is just for rooms we've purged I'm tempted to skip the second clause, as we know we'll be deleting all chains in the room anyway :/

Reply:
I was nervous that we *don't know* it is only for rooms we've purged (if someone has an event missing for some other reason), but I guess we could always re-run this in the future if we find people in that situation.

Reply:
After some discussion we included it in this PR.